### PR TITLE
Add logout feature

### PR DESCRIPTION
**Pull request by emlun**
Meta-PR into https://github.com/syncthing/syncthing/pull/8757. Adds a way to log out.

See: https://github.com/syncthing/syncthing/pull/8757#issuecomment-1519165958

### CODE REVIEW

**Comment by emlun** [comment]
@acolomb @imsodin Thoughts on this? Should we merge this into https://github.com/syncthing/syncthing/pull/8757 ?

**Comment by acolomb** [comment]
Haven't had time to test it yet, but this looks like a reasonable addition. I'd merge it into the other PR.

**Comment by imsodin** [comment]
Personally I'd rather have this as a separate PR in syncthing. There's the principle aspect (that I don't care strongly about here), that it's a new feature/behaviour of it's own compared to the login page. And the more practical aspect: That PR is already large, and I personally have very limited time at the moment for reviewing - so I'd like to focus on getting that in first. However @acolomb and @tomasz1986 are quite actively reviewing, so maybe that's not an issue overall as there's enough "review power" for everthing there.

In any case I'll focus on the tests and main PR for now.

File: lib/api/api.go [hunk 1]
```diff
@@ -363,8 +363,10 @@ func (s *service) Serve(ctx context.Context) error {
 	// Wrap everything in basic auth, if user/password is set.
 	if guiCfg.IsAuthEnabled() {
 		var handlePasswordAuth http.Handler
-		handler, handlePasswordAuth = authAndSessionMiddleware("sessionid-"+s.id.String()[:5], guiCfg, s.cfg.LDAP(), handler, s.evLogger)
+		sessionCookieName := "sessionid-"+s.id.String()[:5]
+		handler, handlePasswordAuth = authAndSessionMiddleware(sessionCookieName, guiCfg, s.cfg.LDAP(), handler, s.evLogger)
 		restMux.Handler(http.MethodPost, "/rest/noauth/auth/password", handlePasswordAuth)
+		restMux.Handler(http.MethodPost, "/rest/noauth/auth/logout", handleLogout(sessionCookieName))
```
Comment (imsodin) [inline lib/api/api.go:369]: Why is this method part of `noauth`?

File: lib/api/api_auth.go [hunk 2]
```diff
@@ -203,6 +205,26 @@ func createSession(cookieName string, username string, guiCfg config.GUIConfigur
...
 }
 
+func handleLogout(cookieName string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		cookie, err := r.Cookie(cookieName)
+		if err == nil && cookie != nil {
+			sessionsMut.Lock()
+			delete(sessions, cookie.Value)
+			sessionsMut.Unlock()
+		}
```
Comment (imsodin) [inline lib/api/api_auth.go:215]: Does it make sense to continue as if it was a success if the condition isn't met (resp. the cookie doesn't exist in sessions)? Maybe related to my question above about this method being part of `noauth`, as if it required auth to call logout the user would get a 403 if the cookie wasn't present (which is what I'd have expected, not saying that's right, just gut feeling).

**Review by imsodin — COMMENTED** [review]
Ok actually after looking at the others circled back here, and it's so small that the practical argument I made isn't that relevant. So whatever suits you best (same or separate PR) is fine with me.

File: lib/api/api.go (hunk 1 above)
Comment (emlun) [reply lib/api/api.go:369]: Hm, good question.

On the one hand, one probably shouldn't be allowed to terminate other users' sessions without their consent (in general principle, ignoring that Syncthing only supports one user anyway...).

On the other hand, if you don't have a session cookie, then the backend won't actually do anything, as the session to be invalidated is determined by the cookie value. And if you do have a valid session cookie, then by definition you are the owner of that session. So it probably doesn't hurt to have this under `/noauth/`?

But yeah, your question indicates that this violates the principle of least surprise. So I guess I'll move it? :slightly_smiling_face:

File: lib/api/api_auth.go (hunk 2 above)
Comment (emlun) [reply lib/api/api_auth.go:215]: I figure if the condition isn't met, then the session has already expired on the server side, so there's no point in keeping the cookie, so you might as well clean it up if the cookie is still there. Either way, from the user's point of view, the result is a successful logout operation, even if there wasn't really anything to do.

As I write this I realize this directly contradicts my conclusion in [the other thread](https://github.com/emlun/syncthing/pull/2/files#r1271457521)... :joy:

Comment (emlun) [reply lib/api/api_auth.go:215]: > Either way, from the user's point of view, the result is a successful logout operation, even if there wasn't really anything to do.

I mean, say you attempt to log out but your session has already expired, so the API hits you with a 403 response. So you'd have to log in again before you can successfully log out... that wouldn't make much sense as a user experience, would it?

File: lib/api/api.go (hunk 1 above)
Comment (imsodin) [reply lib/api/api.go:369]: The least-surprise thing was what triggered my comment in the first place. Maybe there's also value of login/logout being right next to each other. This is one of those things where the longer I think about it the less I have an opinion either way xD

I also had a (anyway very theoretical) attack vector in mind, however writing it out made it disappear: As there's no sleep when one sends an incorrect session id, it's somewhat faster to brute-force it (I know, as I said very theoretical). However if one succeeds, one has the glory of having terminated the session of a user - highly valuable result that is.

File: lib/api/api_auth.go (hunk 2 above)
Comment (imsodin) [reply lib/api/api_auth.go:215]: Yeah, that makes sense.

File: lib/api/api.go (hunk 1 above)
Comment (emlun) [reply lib/api/api.go:369]: Alright, I'll add a comment here but leave it as is. Does that sound fair?

Comment (emlun) [reply lib/api/api.go:369]: I've merged this into https://github.com/syncthing/syncthing/pull/8757, let me know there if you'd like any more changes to this.

**Pull request merged**